shared/js/intl_helper.js

```diff
@@ -0,0 +1,148 @@
+(function(global) {
+  'use strict';
+
+  /*************************
+
+  IntlHelper.defineObject('shortDate', 'datetime', {
+    year: 'numeric',
+    month: 'long',
+  });
+
+  function setValue() {
+    var formatter = IntlHelper.getObject('shortDate');
+    console.log(formatter.format(new Date()));
+  }
+
+  setValue();
+
+  IntlHelper.on('shortDate', setValue);
+  
+   ****************************/
+  var objectCache;
+
+  var knownObjects = [
+    'datetime',
+    'number',
+  ];
+
+  global.IntlHelper = {
+    defineObject: function(name, type, options) {
+      if (knownObjects.indexOf(type) === -1) {
+        throw new Error('Unknown type: ' + type);
+      }
+
+      if (!objectCache) {
+        objectCache = new Map();
+      }
+
+      if (objectCache.has(name)) {
+        throw new Error(
+          'Intl Object with name "' + name + '" already exists');
+      }
+
+      objectCache.set(name, {
+        options,
+        type,
+        object: undefined,
+        listeners: new Set()
+      });
+    },
+
+    getObject: function(name) {
+      if (!objectCache.has(name)) {
+        throw new Error('Intl Object with name "' + name + '" is not defined');
+      }
+      const obj = objectCache.get(name);
+
+      if (obj.object === undefined) {
+        switch (obj.type) {
+          case 'number':
+            obj.object = Intl.NumberFormat(
+              navigator.languages,
+              obj.options
+            );
+            break;
+          case 'datetime':
+            const options = Object.assign({}, obj.options);
+            if (options.hour) {
+              options.hour12 = navigator.mozHour12;
+            }
+            console.log('new Format');
+            console.log(options);
+            obj.object = Intl.DateTimeFormat(
+              navigator.languages,
+              options
+            );
+            break;
+          default:
+            throw new Error('Unknown type: ' + obj.type);
+        }
+      }
+      return obj.object;
+    },
+    on: function(names, cb) {
+      if (typeof names === 'string') {
+        names = [names];
```

Comment:
So, the reason I went for names, is that when you write a function that handles reformatting, the function covers specific set of formatters, not formatter types.

What I mean by this is that if you imagine using IntlHelper in SMS, for example, you may have a function that formats date/time in Conversation view.
That function will handle one or two DateTimeFormatters with specific names.

Now, if IntlHelper.on takes `type` then it will fire your reformatting function whenever *any* datetimeformat changes. If it takes a list of names, it will fire only when one of the affected datetimeformats change. It seems to me that desigining the API around names allows me to be smarter in the future.

One example use case I see is where there are two datetimeformatters, one with month/year/day and another with hour/minute. When timeformatchanges, I don't need to redefine the former one, so any callback that uses only the former one doesn't have to be fired at all - because month/year/day is not affected. But any callback that covers the latter should be - hour is affected.

If I follow 'types' then I don't have this kind of control. I have to fire all callbacks and trigger all reformatting every time any datetimeformat changes.